### DISCUSSION TITLE
BIGTREETECH SFS V2.0 Smart Filament Runout Sensor

### DISCUSSION THREAD

**Discussion by johncotten** [xblax/flashforge_ad5m_klipper_mod#276] [Development]
If you aren't familiar with this sensor, it not only detects filament runout but it also detects clogs and jamming. 

I was an early adopter on v.03 primarily to update the crappy runout sensor with a smarter sensor. The Non Klipperscreen version V.03 has been working fantastic all this time. But I occasionally had a glitch where the "pause" command would only temporarily home then start printing 20 mm off the surface. This was not this mod problem but something in the 2.4.5 firmware I was running. 
Today I decided to upgrade to the latest 2.7.8 and v.05 of this mod. Everything except wifi seems to work on the guppy screen version. There was a few changes I needed to do, and decided I would share for anyone wanting to update to this Runout Sensor.

In the macros.cfg you will need to change the variable_runout_sensor in _CLIENT_VARIABLE to
variable_runout_sensor    : "filament_motion_sensor runout_sensor"

I also changed these because it was parking to change out filament in the front left which wasn't possible to do:
variable_custom_park_x    : 105.0  ; custom x position; value must be within your defined min and max of X
variable_custom_park_y    : 105.0  ; custom y position; value must be within your defined min and max of Y
variable_custom_park_dz   : 30.0    ; custom dz value; the value in mm to lift the nozzle when move to park position

In the printer.base.cfg I changed:
# Filament runout sensor
[filament_switch_sensor runout_sensor]
pause_on_runout: False
switch_pin: !PB14
event_delay: 1.0
runout_gcode:
    _FILAMENT_RUNOUT_EVENT

to 

[filament_motion_sensor runout_sensor]
pause_on_runout: False
switch_pin: !PB14
detection_length: 5.0
extruder: extruder
runout_gcode:
    _FILAMENT_RUNOUT_EVENT
    M117 Filament runout

The cable for the BTT is a 4 pin cable that is split. You have to rein the cable to match the below image. Red,Green,Back with the blue not connected to anything.  
![IMG_6745](https://github.com/user-attachments/assets/7849caff-6487-4c36-a18a-b5e97d7b697c)

**Comment by mallison01** [comment]
Hey how did you mount the sensor, I have the diy enclosure and here lately my sensor has been getting some false triggers, so I was thinking of upgrading it to the biqu.

**Reply by johncotten** [reply]
Sorry for the late reply. I just used sticky tape on the outside. Hope that helps. Has been holding at least since last March.
![IMG_6742](https://github.com/user-attachments/assets/dabec740-5a6b-4d68-858a-153c9c931e56)
![IMG_6746](https://github.com/user-attachments/assets/e9b76266-3034-4098-80b2-01ec1c29148d)

**Comment by katachasm-ux** [comment]
I keep getting filament cracks in the PTFE tube *after* the sensor, is there any way to mount this on the print head?

**Comment by wenli767** [comment]
I'm adding on because you are the only other person who installed the BTT SFS V2 on your AD5M printer. I cut the old 3 pin cable and spliced it with the BTT's 4 pin cable in the same order as you. Didn't work for me, the light was solid blue at all times never changing to red or purple. Then I realized that the old 3 pin cable has white tape wrapped at the end marking "Silk". And under that tape the top cable was actually the middle cable. After switching that and making it actually match what you have (red-green-black) then everything worked. So if anyone else have a similar issue then double check under the tape.
